README
# a C10K server taking advantage of memcached's network library and websocket protocol
# 一个由memcached的网络库和websocket协议构成的长连接服务器

## Levin Add websoket support to memcached
* vllm@163.com

[USAGE]
    [.INSTALATION]
        [..1]install libevent
        [..2]./configure --with-libevent=/path/to/libevent/install/dir
        [..3]make
        [..4]make install
        [..5]/usr/local/bin/memcached -u root
    [.TEST WEBSOCK HANDSHAKE]
        [..6]open another terminal window
        [..7]connect server with websocket handshake header by client(written in PHP or anything you like)
        [..8]chek and enjoy it
        
[FUNTION]
    [.DONE]
        [..1]support websoket handshake
    [.TODO]
        [..2]support websoket full protocol
        [..3]code optimize 

## Dependencies

* libevent, http://www.monkey.org/~provos/libevent/ (libevent-dev)

## Environment

### Linux

If using Linux, you need a kernel with epoll.  Sure, libevent will
work with normal select, but it sucks.

epoll isn't in Linux 2.4, but there's a backport at:

    http://www.xmailserver.org/linux-patches/nio-improve.html

You want the epoll-lt patch (level-triggered).

## Website

* http://www.memcached.org

## Contributing

Want to contribute?  Up-to-date pointers should be at:

* http://contributing.appspot.com/memcached


ChangeLog
2015-04-30
        * first comment by levin

2010-10-11
        * ChangeLog is no longer being updated.
        See http://code.google.com/p/memcached/wiki/ReleaseNotes

2009-04-10

        * *critical bugfix*. In 1.2.7 under multithreaded mode, memcached
          would never restart accepting connections after hitting the
          maximum connection limit.

2009-04-02

        * Fix: Rewrote stat buffer handling (Trond)

2009-03-31

        * Feature:  -b option for configuring backlog

        * Feature:  stats settings

        * Fix: binary stat alignment bug (bug26)

        * Fix: buffer overflow in stats (bug27)

        * Fix: recycle memory more aggressively (bug14)

        * Fix: incr validation (bug31)

        * Fix: 64-bit incr/decr delta problem (bug21)

        * Fix: Ascii UDP set (bug36)

        * Fix: stats slabs' used chunks (bug29)

        * Fix: stats reset resetting more (bug22)

        * Misc: More tests, documentation, cleanliness (godliness)

        * Stable merge (stats, debuggability, listen bugs)

2009-03-11

        * Protocol:  Binary complete (Dustin, Trond, Toru, etc...)

        * Performance:  improvements from facebook (Dormando)

        * Performance:  per-thread stats (Trond)

        * Performance:  Hash expansion in its own thread (Trond)

        * Stats:  Lots of new stats (Dustin)

        * Bug fixes (various)

        See the 1.3.2 release notes or revision control for more detail.

2008-09-06
       * Display error status on listen failures (Dormando)

       * Remove managed instance code. Incomplete/etc. (Dormando)

          * Handle broken IPV6 stacks better (Brian Aker)

          * Generate warnings on setsockopt() failures (Brian Aker)

          * Fix some indentation issues (Brian Aker)

          * UDP/TCP can be disabled by setting their port to zero (Brian Aker)

          * Zero out libevent thread structures before use (Ricky Zhou)

          * New stat: Last accessed time for last evicted item per slab class.
            (Dormando)

          * Use a dedicated socket accept thread (Facebook)

          * Add -R option. Limit the number of requests processed by a connection
            at once. Prevents starving other threads if bulk loading. (Facebook)

2008-07-29 [Version 1.2.6 released]

2008-07-24 [Version 1.2.6-rc1 released]

       * Add support for newer automake (Facebook)

       * DTrace support for Solaris/etc (Trond Norbye)

       * LRU tests (Steve Yen)

       * Handle negative length items properly (Dormando)

       * Don't leave stale data after failed set attempts (Dormando)

       * Fix refcount leaks, which would result in OOM's on all sets
               (Dormando)

       * Fix buffer overruns (Dustin Sallings, Tomash Brechko)

       * Fix memory corruption with CAS (Dustin Sallings)

2008-06-11

       * Fix -k to work with -d. (reported by Gary Zhu)

2008-03-02 [Version 1.2.5-rc1 released]

       * Add per-item-class tracking of evictions and OOM errors (dormando)

       * Optimize item_alloc() a little (dormando)

       * Give 'SERVER_ERROR out of memory' errors more context (dormando)

       * Enable usage of large memory pages under solaris
         (Trond.Norbye@Sun.COM)

       * Enable UDP by default, clean up server socket code
         (brian@tangent.org)

       * 'noreply' support (Tomash Brechko)

       * IPv6 support, and IPv6 multi-interface support (brian@tangent.org)

       * Add compiler options for Sun Studio compilers with --enable-threads
             (Trond.Norbye@Sun.COM)

       * Add --enable-64bit for mulitarget platforms (Trond.Norbye@Sun.COM)

       * Use gettimeofday(2) instead of time(2).

       * Make -k option work (Tomash Brechko)

           * Fix chunk slab alignment (Trond.Norbye@Sun.COM)

2007-12-06 [Version 1.2.4 released]

2007-12-05

       * Fix compilation on panther (JS and Dormando)

           * More CAS tests (Chris Goffinet)

           * Final fixes for all 1.2.4 features are in, -rc2 sent out.

2007-11-19 [Version 1.2.4-rc1 released]

2007-11-19  Dormando <dormando@rydia.net>

       * Patch series from Tomash Brechko <tomash.brechko@gmail.com>:
         Minor fixes and optimisations.

       * Patches from Chris, Dustin, and Dormando to fix CAS.

       * Prepping for 1.2.4 release.

2007-11-13  Dormando <dormando@rydia.net>

        * Adjusted patch from js <ebgssth@gmail.com>: Compile on OS X Panther
          and earlier.

2007-11-12  Steven Grimm  <sgrimm@facebook.com>

        * Patch from Tomash Brechko <tomash.brechko@gmail.com>: Always send
          "SERVER_ERROR out of memory" when memory exhausted.

2007-10-15  Paul Lindner  <lindner@inuus.com>

        * Patch from David Bremner <bremner@unb.ca> that implements
          a new option "-a" which takes an octal permission mask
          (like chmod) sets the permissions on the unix domain socket 
          (specified by "-s").

2007-10-03 Paul Lindner <lindner@inuus.com>
        * Incorporate "cas" operation developed by Dustin
          Sallings <dustin@spy.net> This change allows you
          to do atomic changes to an existing key.

        * Fix for stats.evictions not incrementing
          when exptime == 0 items are kicked off the cache. 
          from Jean-Francois BUSTARRET <jfbustarret@wat.tv>.

        * Fix for do_item_cachedump() which was returning
          an incorrect timestamp.

        * Switch to unsigned 64-bit increment/decrement counters
          from Evan Miller and Dustin Sallings.

        * Add append command support written by Filipe Laborde.
          Thread safe version plus prepend command from Maxim Dounin
          <mdounin@mdounin.ru>

        * The memcached-tool script can now display stats.  Patch
          provided by Dan Christian <dchristian@google.com>

        * Fix for Unix Domain sockets on FreeBSD
          FreeBSD's sendmsg() requires msg_name in msghdr structure 

base64.h
/* base64.h -- Encode binary data using printable characters.
   Copyright (C) 2004, 2005, 2006 Free Software Foundation, Inc.
   Written by Simon Josefsson.
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */

#ifndef BASE64_H
# define BASE64_H

/* Get size_t. */
# include <stddef.h>

/* Get bool. */
# include <stdbool.h>

/* This uses that the expression (n+(k-1))/k means the smallest
   integer >= n/k, i.e., the ceiling of n/k.  */
# define BASE64_LENGTH(inlen) ((((inlen) + 2) / 3) * 4)

extern bool isbase64 (char ch);

extern void base64_encode (const char *in, size_t inlen,
			   char *out, size_t outlen);

extern size_t base64_encode_alloc (const char *in, size_t inlen, char **out);

extern bool base64_decode (const char *in, size_t inlen,
			   char *out, size_t *outlen);

extern bool base64_decode_alloc (const char *in, size_t inlen,
				 char **out, size_t *outlen);

#endif /* BASE64_H */


config.h
/* config.h.  Generated by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Set to nonzero if you want to include DTRACE */
/* #undef ENABLE_DTRACE */

/* Set to nonzero if you want to include SASL */
/* #undef ENABLE_SASL */

/* Set to nonzero if you want to enable a SASL pwdb */
/* #undef ENABLE_SASL_PWDB */

/* machine is bigendian */
/* #undef ENDIAN_BIG */

/* machine is littleendian */
#define ENDIAN_LITTLE 1

/* Define to 1 if support accept4 */
/* #undef HAVE_ACCEPT4 */

/* Define to 1 if you have the `clock_gettime' function. */
#define HAVE_CLOCK_GETTIME 1

/* Define this if you have an implementation of drop_privileges() */
/* #undef HAVE_DROP_PRIVILEGES */

/* GCC Atomics available */
#define HAVE_GCC_ATOMICS 1

/* Define to 1 if you have the `getpagesizes' function. */
/* #undef HAVE_GETPAGESIZES */

/* Have ntohll */
/* #undef HAVE_HTONLL */

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the `memcntl' function. */
/* #undef HAVE_MEMCNTL */

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the `mlockall' function. */
#define HAVE_MLOCKALL 1

/* we have sasl_callback_ft */
/* #undef HAVE_SASL_CALLBACK_FT */

/* Set to nonzero if your SASL implementation supports SASL_CB_GETCONF */
/* #undef HAVE_SASL_CB_GETCONF */

/* Define to 1 if you have the <sasl/sasl.h> header file. */
#define HAVE_SASL_SASL_H 1

/* Define to 1 if you have the `setppriv' function. */
/* #undef HAVE_SETPPRIV */

/* Define to 1 if you have the `sigignore' function. */
#define HAVE_SIGIGNORE 1

/* Define to 1 if stdbool.h conforms to C99. */
#define HAVE_STDBOOL_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define this if you have umem.h */
/* #undef HAVE_UMEM_H */

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if the system has the type `_Bool'. */
#define HAVE__BOOL 1

/* Machine need alignment */
/* #undef NEED_ALIGN */

/* Define to 1 if your C compiler doesn't accept -c and -o together. */
/* #undef NO_MINUS_C_MINUS_O */

/* Name of package */
#define PACKAGE "memcached"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "linliming@baidu.com"

/* Define to the full name of this package. */
#define PACKAGE_NAME "memcached"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "memcached 1.4.22"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "memcached"

/* Define to the version of this package. */
#define PACKAGE_VERSION "1.4.22"

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "1.4.22"

/* find sigignore on Linux */
#define _GNU_SOURCE 1

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* define to int if socklen_t not available */
/* #undef socklen_t */

#if HAVE_STDBOOL_H
#include <stdbool.h>
#else
#define bool char
#define false 0
#define true 1
#endif

#ifdef HAVE_INTTYPES_H
#include <inttypes.h>
#endif



base64.c
/* base64.c -- Encode binary data using printable characters.
   Copyright (C) 1999, 2000, 2001, 2004, 2005, 2006 Free Software
   Foundation, Inc.
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */

/* Written by Simon Josefsson.  Partially adapted from GNU MailUtils
 * (mailbox/filter_trans.c, as of 2004-11-28).  Improved by review
 * from Paul Eggert, Bruno Haible, and Stepan Kasal.
 *
 * See also RFC 3548 <http://www.ietf.org/rfc/rfc3548.txt>.
 *
 * Be careful with error checking.  Here is how you would typically
 * use these functions:
 *
 * bool ok = base64_decode_alloc (in, inlen, &out, &outlen);
 * if (!ok)
 *   FAIL: input was not valid base64
 * if (out == NULL)
 *   FAIL: memory allocation error
 * OK: data in OUT/OUTLEN
 *
 * size_t outlen = base64_encode_alloc (in, inlen, &out);
 * if (out == NULL && outlen == 0 && inlen != 0)
 *   FAIL: input too long
 * if (out == NULL)
 *   FAIL: memory allocation error
 * OK: data in OUT/OUTLEN.
 *
 */

/* Get prototype. */
#include "base64.h"

/* Get malloc. */
#include <stdlib.h>

/* Get UCHAR_MAX. */
#include <limits.h>

/* C89 compliant way to cast 'char' to 'unsigned char'. */
static inline unsigned char to_uchar (char ch) {
  return ch;
}

/* Base64 encode IN array of size INLEN into OUT array of size OUTLEN.
   If OUTLEN is less than BASE64_LENGTH(INLEN), write as many bytes as
   possible.  If OUTLEN is larger than BASE64_LENGTH(INLEN), also zero
   terminate the output buffer. */
void base64_encode (const char *in, size_t inlen, char *out, size_t outlen) {
  static const char b64str[64] =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

  while (inlen && outlen)
    {
      *out++ = b64str[(to_uchar (in[0]) >> 2) & 0x3f];
      if (!--outlen)
	break;
      *out++ = b64str[((to_uchar (in[0]) << 4)
		       + (--inlen ? to_uchar (in[1]) >> 4 : 0))
		      & 0x3f];
      if (!--outlen)
	break;
      *out++ =
	(inlen
	 ? b64str[((to_uchar (in[1]) << 2)
		   + (--inlen ? to_uchar (in[2]) >> 6 : 0))
		  & 0x3f]
	 : '=');
      if (!--outlen)
	break;
      *out++ = inlen ? b64str[to_uchar (in[2]) & 0x3f] : '=';
      if (!--outlen)
	break;
      if (inlen)
	inlen--;
      if (inlen)
	in += 3;
    }

  if (outlen)
    *out = '\0';
}

/* Allocate a buffer and store zero terminated base64 encoded data
   from array IN of size INLEN, returning BASE64_LENGTH(INLEN), i.e.,
   the length of the encoded data, excluding the terminating zero.  On
   return, the OUT variable will hold a pointer to newly allocated
   memory that must be deallocated by the caller.  If output string
   length would overflow, 0 is returned and OUT is set to NULL.  If
   memory allocation failed, OUT is set to NULL, and the return value
   indicates length of the requested memory block, i.e.,
   BASE64_LENGTH(inlen) + 1. */
size_t base64_encode_alloc (const char *in, size_t inlen, char **out) {
  size_t outlen = 1 + BASE64_LENGTH (inlen);

  /* Check for overflow in outlen computation.
   *
   * If there is no overflow, outlen >= inlen.
   *
   * If the operation (inlen + 2) overflows then it yields at most +1, so
   * outlen is 0.
   *
   * If the multiplication overflows, we lose at least half of the
   * correct value, so the result is < ((inlen + 2) / 3) * 2, which is
   * less than (inlen + 2) * 0.66667, which is less than inlen as soon as
   * (inlen > 4).
   */
  if (inlen > outlen)
    {
      *out = NULL;
      return 0;
    }

  *out = malloc (outlen);
  if (!*out)
    return outlen;

  base64_encode (in, inlen, *out, outlen);

  return outlen - 1;
}

/* With this approach this file works independent of the charset used
   (think EBCDIC).  However, it does assume that the characters in the
   Base64 alphabet (A-Za-z0-9+/) are encoded in 0..255.  POSIX
   1003.1-2001 require that char and unsigned char are 8-bit
   quantities, though, taking care of that problem.  But this may be a
   potential problem on non-POSIX C99 platforms.
   IBM C V6 for AIX mishandles "#define B64(x) ...'x'...", so use "_"
   as the formal parameter rather than "x".  */
#define B64(_)					\
  ((_) == 'A' ? 0				\
   : (_) == 'B' ? 1				\
   : (_) == 'C' ? 2				\
   : (_) == 'D' ? 3				\
   : (_) == 'E' ? 4				\
   : (_) == 'F' ? 5				\
   : (_) == 'G' ? 6				\
   : (_) == 'H' ? 7				\
   : (_) == 'I' ? 8				\
   : (_) == 'J' ? 9				\
   : (_) == 'K' ? 10				\
   : (_) == 'L' ? 11				\
   : (_) == 'M' ? 12				\
   : (_) == 'N' ? 13				\
   : (_) == 'O' ? 14				\
   : (_) == 'P' ? 15				\
   : (_) == 'Q' ? 16				\
   : (_) == 'R' ? 17				\
   : (_) == 'S' ? 18				\
   : (_) == 'T' ? 19				\
   : (_) == 'U' ? 20				\
   : (_) == 'V' ? 21				\
   : (_) == 'W' ? 22				\
   : (_) == 'X' ? 23				\
   : (_) == 'Y' ? 24				\
   : (_) == 'Z' ? 25				\
   : (_) == 'a' ? 26				\
   : (_) == 'b' ? 27				\
   : (_) == 'c' ? 28				\
   : (_) == 'd' ? 29				\
   : (_) == 'e' ? 30				\
   : (_) == 'f' ? 31				\
   : (_) == 'g' ? 32				\
   : (_) == 'h' ? 33				\
   : (_) == 'i' ? 34				\
   : (_) == 'j' ? 35				\
   : (_) == 'k' ? 36				\
   : (_) == 'l' ? 37				\
   : (_) == 'm' ? 38				\
   : (_) == 'n' ? 39				\
   : (_) == 'o' ? 40				\
   : (_) == 'p' ? 41				\
   : (_) == 'q' ? 42				\
   : (_) == 'r' ? 43				\
   : (_) == 's' ? 44				\
   : (_) == 't' ? 45				\
   : (_) == 'u' ? 46				\
   : (_) == 'v' ? 47				\
   : (_) == 'w' ? 48				\
   : (_) == 'x' ? 49				\
   : (_) == 'y' ? 50				\
   : (_) == 'z' ? 51				\
   : (_) == '0' ? 52				\
   : (_) == '1' ? 53				\
   : (_) == '2' ? 54				\
   : (_) == '3' ? 55				\
   : (_) == '4' ? 56				\
   : (_) == '5' ? 57				\
   : (_) == '6' ? 58				\
   : (_) == '7' ? 59				\
   : (_) == '8' ? 60				\
   : (_) == '9' ? 61				\
   : (_) == '+' ? 62				\
   : (_) == '/' ? 63				\
   : -1)

static const signed char b64[0x100] = {
  B64 (0), B64 (1), B64 (2), B64 (3),
  B64 (4), B64 (5), B64 (6), B64 (7),
  B64 (8), B64 (9), B64 (10), B64 (11),
  B64 (12), B64 (13), B64 (14), B64 (15),
  B64 (16), B64 (17), B64 (18), B64 (19),
  B64 (20), B64 (21), B64 (22), B64 (23),
  B64 (24), B64 (25), B64 (26), B64 (27),
  B64 (28), B64 (29), B64 (30), B64 (31),
  B64 (32), B64 (33), B64 (34), B64 (35),
  B64 (36), B64 (37), B64 (38), B64 (39),
  B64 (40), B64 (41), B64 (42), B64 (43),
  B64 (44), B64 (45), B64 (46), B64 (47),
  B64 (48), B64 (49), B64 (50), B64 (51),
  B64 (52), B64 (53), B64 (54), B64 (55),
  B64 (56), B64 (57), B64 (58), B64 (59),
  B64 (60), B64 (61), B64 (62), B64 (63),
  B64 (64), B64 (65), B64 (66), B64 (67),
  B64 (68), B64 (69), B64 (70), B64 (71),
  B64 (72), B64 (73), B64 (74), B64 (75),
  B64 (76), B64 (77), B64 (78), B64 (79),
  B64 (80), B64 (81), B64 (82), B64 (83),
  B64 (84), B64 (85), B64 (86), B64 (87),
  B64 (88), B64 (89), B64 (90), B64 (91),
  B64 (92), B64 (93), B64 (94), B64 (95),
  B64 (96), B64 (97), B64 (98), B64 (99),
  B64 (100), B64 (101), B64 (102), B64 (103),
  B64 (104), B64 (105), B64 (106), B64 (107),
  B64 (108), B64 (109), B64 (110), B64 (111),
  B64 (112), B64 (113), B64 (114), B64 (115),
  B64 (116), B64 (117), B64 (118), B64 (119),
  B64 (120), B64 (121), B64 (122), B64 (123),
  B64 (124), B64 (125), B64 (126), B64 (127),
  B64 (128), B64 (129), B64 (130), B64 (131),
  B64 (132), B64 (133), B64 (134), B64 (135),
  B64 (136), B64 (137), B64 (138), B64 (139),
  B64 (140), B64 (141), B64 (142), B64 (143),
  B64 (144), B64 (145), B64 (146), B64 (147),
  B64 (148), B64 (149), B64 (150), B64 (151),
  B64 (152), B64 (153), B64 (154), B64 (155),
  B64 (156), B64 (157), B64 (158), B64 (159),
  B64 (160), B64 (161), B64 (162), B64 (163),
  B64 (164), B64 (165), B64 (166), B64 (167),
  B64 (168), B64 (169), B64 (170), B64 (171),
  B64 (172), B64 (173), B64 (174), B64 (175),
  B64 (176), B64 (177), B64 (178), B64 (179),
  B64 (180), B64 (181), B64 (182), B64 (183),
  B64 (184), B64 (185), B64 (186), B64 (187),
  B64 (188), B64 (189), B64 (190), B64 (191),
  B64 (192), B64 (193), B64 (194), B64 (195),
  B64 (196), B64 (197), B64 (198), B64 (199),
  B64 (200), B64 (201), B64 (202), B64 (203),
  B64 (204), B64 (205), B64 (206), B64 (207),
  B64 (208), B64 (209), B64 (210), B64 (211),
  B64 (212), B64 (213), B64 (214), B64 (215),
  B64 (216), B64 (217), B64 (218), B64 (219),
  B64 (220), B64 (221), B64 (222), B64 (223),
  B64 (224), B64 (225), B64 (226), B64 (227),
  B64 (228), B64 (229), B64 (230), B64 (231),
  B64 (232), B64 (233), B64 (234), B64 (235),
  B64 (236), B64 (237), B64 (238), B64 (239),
  B64 (240), B64 (241), B64 (242), B64 (243),
  B64 (244), B64 (245), B64 (246), B64 (247),
  B64 (248), B64 (249), B64 (250), B64 (251),
  B64 (252), B64 (253), B64 (254), B64 (255)
};

#if UCHAR_MAX == 255
# define uchar_in_range(c) true
#else
# define uchar_in_range(c) ((c) <= 255)
#endif

/* Return true if CH is a character from the Base64 alphabet, and
   false otherwise.  Note that '=' is padding and not considered to be
   part of the alphabet.  */
bool isbase64 (char ch) {
  return uchar_in_range (to_uchar (ch)) && 0 <= b64[to_uchar (ch)];
}

/* Decode base64 encoded input array IN of length INLEN to output
   array OUT that can hold *OUTLEN bytes.  Return true if decoding was
   successful, i.e. if the input was valid base64 data, false
   otherwise.  If *OUTLEN is too small, as many bytes as possible will
   be written to OUT.  On return, *OUTLEN holds the length of decoded
   bytes in OUT.  Note that as soon as any non-alphabet characters are
   encountered, decoding is stopped and false is returned.  This means
   that, when applicable, you must remove any line terminators that is
   part of the data stream before calling this function.  */
bool base64_decode (const char *in, size_t inlen, char *out, size_t *outlen) {
  size_t outleft = *outlen;

  while (inlen >= 2)
    {
      if (!isbase64 (in[0]) || !isbase64 (in[1]))
	break;

      if (outleft)
	{
	  *out++ = ((b64[to_uchar (in[0])] << 2)
		    | (b64[to_uchar (in[1])] >> 4));
	  outleft--;
	}

      if (inlen == 2)
	break;

      if (in[2] == '=')
	{
	  if (inlen != 4)
	    break;

	  if (in[3] != '=')
	    break;

	}
      else
	{
	  if (!isbase64 (in[2]))
	    break;

	  if (outleft)
	    {
	      *out++ = (((b64[to_uchar (in[1])] << 4) & 0xf0)
			| (b64[to_uchar (in[2])] >> 2));
	      outleft--;
	    }

	  if (inlen == 3)
	    break;

	  if (in[3] == '=')
	    {
	      if (inlen != 4)
		break;
	    }
	  else
	    {
	      if (!isbase64 (in[3]))
		break;

	      if (outleft)
		{
		  *out++ = (((b64[to_uchar (in[2])] << 6) & 0xc0)
			    | b64[to_uchar (in[3])]);
		  outleft--;
		}
	    }
	}

      in += 4;
      inlen -= 4;
    }

  *outlen -= outleft;

  if (inlen != 0)
    return false;

  return true;
}

/* Allocate an output buffer in *OUT, and decode the base64 encoded
   data stored in IN of size INLEN to the *OUT buffer.  On return, the
   size of the decoded data is stored in *OUTLEN.  OUTLEN may be NULL,
   if the caller is not interested in the decoded length.  *OUT may be
   NULL to indicate an out of memory error, in which case *OUTLEN
   contains the size of the memory block needed.  The function returns
   true on successful decoding and memory allocation errors.  (Use the
   *OUT and *OUTLEN parameters to differentiate between successful
   decoding and memory error.)  The function returns false if the
   input was invalid, in which case *OUT is NULL and *OUTLEN is
   undefined. */
bool
base64_decode_alloc (const char *in, size_t inlen, char **out,
		     size_t *outlen)
{
  /* This may allocate a few bytes too much, depending on input,
     but it's not worth the extra CPU time to compute the exact amount.
     The exact amount is 3 * inlen / 4, minus 1 if the input ends
     with "=" and minus another 1 if the input ends with "==".
     Dividing before multiplying avoids the possibility of overflow.  */
  size_t needlen = 3 * (inlen / 4) + 2;

  *out = malloc (needlen);
  if (!*out)
    return true;

  if (!base64_decode (in, inlen, *out, &needlen))
    {
      free (*out);
      *out = NULL;
      return false;
    }

  if (outlen)
    *outlen = needlen;

  return true;
}
